### PR TITLE
Plugin: Implement block registering API

### PR DESCRIPTION
**Pull request by youknowriad**
**Questions**

 - I was thinking about adding unit tests for these fonctions, but setting up QUnit is not an easy task with our setup. Do you think we could use `mocha` maybe?
 - Also: Using a library like `lodash` here seems appropriate, what are your thoughts on adding it?
 - We're using a tuple (namespace, name) to identify a block, should we instead merge them into one `id` ( `namespace/name` )?

### CODE REVIEW

File: external/@wordpress/blocks/index.js
Comment (youknowriad) [inline external/@wordpress/blocks/index.js]: It's not keyed by block name because the block name is not unique, it's unique inside a namespace only.

Comment (aduth) [inline external/@wordpress/blocks/index.js]: Yeah, I'm worried we'll find the namespacing will get in our way more than it helps. I'm not totally convinced about it. It's nice to have the consistency with similar APIs for the REST API, but I wonder if we could just allow a simple name and perhaps encourage a convention around including some namespace in the name itself: like `"wp/text"` (or without namespace for "core" types).

File: external/@wordpress/blocks/index.js
Comment (aduth) [inline external/@wordpress/blocks/index.js]: If we call it `name` on the object, should we just accept the argument labeled as `name` as well?

File: external/@wordpress/blocks/index.js
Comment (aduth) [inline external/@wordpress/blocks/index.js]: Might be nice to provide a DocBlock for this.

File: external/@wordpress/blocks/index.js
Comment (nylen) [inline external/@wordpress/blocks/index.js]: What are the expected uses of this function?  Do you think we will need the ability to pull a specific block out of the list?  If so, it might be a good idea to return an array keyed by namespace then block name, as the blocks are stored.

Also, `Block settings` is better wording.

Comment (aduth) [inline external/@wordpress/blocks/index.js]: >What are the expected uses of this function?

This is a good point. Seems nice as an external API, but I don't see an obvious use-case for us initially.

Comment (youknowriad) [inline external/@wordpress/blocks/index.js]: From my experience with the `multi-instance` prototype. This function was used in two places:

 - Fetching all the blocks to show in the block inserter (Array is better here, because we need to group them by category and not by namespace)

 - Fetching all the blocks to find the blocks that a block can be switched to (The filtering happens using one of the settings key and not the namespace).

So IMO it's better to return a simple array.

Regarding the name, I'm leaning towards removing the namespace and encouraging a convention

**Comment by nylen** [comment]
> I was thinking about adding unit tests for these functions [...] Do you think we could use mocha maybe?

Yes, let's go ahead and do this please :)  Whatever we choose should support browser testing as well.

I agree I like Mocha better, but core uses QUnit.  Do you see this causing problems in the future if we need to integrate both?  What are the difficulties in adding QUnit?

> Also: Using a library like lodash here seems appropriate, what are your thoughts on adding it?

+1.  WP core already bundles Underscore, though Lodash is nicer to work with and we would be able to include selective imports in a build process.

> We're using a tuple (namespace, name) to identify a block, should we instead merge them into one id ( namespace/name )?

If we do keep the namespace and name separate, this makes it easier to test and enforce guidelines for block authors using these fields correctly.

**Comment by youknowriad** [comment]
> If we do keep the namespace and name separate

The problem with keeping the name and namespace separate is that each time we'll need to retrieve the block settings we'll do `getBlockSettings( blockState.name, blockState.namespace)` which means we need to store a namespace and a name in the block state and in the generated markup (or metadata). `<!-- wp:text namespace:core -->` while we could keep it simpler `<!-- wp:core/text -->`

File: external/@wordpress/blocks/index.js
Comment (nylen) [inline external/@wordpress/blocks/index.js]: > Regarding the name, I'm leaning towards removing the namespace and encouraging a convention

_Encouraging_ a convention is not enough, it needs to be _enforced_ otherwise people won't follow it.

Though this can be done either way -

```js
if ( ! /^\w+/\w+$/.test( blockName ) ) {
    throw new Error( 'Use a namespace please' );
}
```

**Comment by nylen** [comment]
> getBlockSettings( blockState.name, blockState.namespace )

Good point, that is not ideal :)

As long as we can do something like https://github.com/WordPress/gutenberg/pull/289#discussion_r106672397, then combining the namespace and name makes sense to me.

**Comment by youknowriad** [comment]
I updated the API and dropped the namespace.

>  Do you see this causing problems in the future if we need to integrate both? What are the difficulties in adding QUnit?

My experience with QUnit is equal to Zero :), So after reading the docs. I figured out that integrating it with our build system means having a separate webpack.config for the tests and import all the test files on a test entry point file. Not so handy IMO, but maybe I'm missing a better way to setup qunit. If someone could help on setting up QUnit (could be a separate PR), it would be awesome!!!

**Comment by nylen** [comment]
I would very much prefer that we start now with the rule of "do not merge code unless it has tests".

Also, mocha is going to be much nicer to work with than QUnit.  I'll try to add it here.

File: external/@wordpress/blocks/index.js
Comment (mtias) [inline external/@wordpress/blocks/index.js]: Repeating my comment elsewhere, I think we can let go of namespace. The name should be unique and descriptive anyways because the user would interact and search for it.

**Comment by nylen** [comment]
@mtias I'm replying to your https://github.com/WordPress/gutenberg/pull/289#discussion_r106701134 here because the thread has disappeared after a rebase.

> Repeating my comment elsewhere, I think we can let go of namespace. The name should be unique and descriptive anyways because the user would interact and search for it.

I am not sure whether you mean letting go of namespace as a separate argument, or the concept entirely.  I agree that using the namespace as a separate argument is going to lead to more complicated code, removing it in 65d60a6 has already simplified things a good bit.

But we really need to require that blocks use a plugin-specific prefix for their internal ID/name/slug/whatever you call it, and in general we need to start thinking about how blocks can be named and how their resulting `<!-- wp -->` tags will look.  Otherwise there will be collisions and it will be unpleasant.

There is a lot of precedent in WP core for doing similar things.  The reasoning is laid out really well in this blog post:  https://nacin.com/2010/05/11/in-wordpress-prefix-everything/

Here are a couple more specific examples:

- https://developer.wordpress.org/plugins/the-basics/best-practices/
- https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#namespacing

IMO, it's also worth considering whether we need to version blocks similarly to REST API endpoints.  This is something we could add later on if we need it, though.

**Comment by mtias** [comment]
@nylen yes, sorry for the confusion, I mean using specific argument or `/` syntax for namespacing and just doing `myplugin-map` or `myplugin_map`.

**Comment by nylen** [comment]
I'm not planning to spend a lot of time on it, beyond making it work and writing some simple tests, but enforcing a very specific syntax is a good thing.  We can always relax the rules later when more needs become apparent, but we can't do the reverse when people misuse the functionality provided.

File: external/@wordpress/blocks/index.js
Comment (aduth) [inline external/@wordpress/blocks/index.js]: This will probably require pulling in [`babel-plugin-transform-runtime`](https://www.npmjs.com/package/babel-plugin-transform-runtime), otherwise will error on browsers where this is not yet implemented.

File: .babelrc
Comment (aduth) [inline .babelrc]: This was important to have in order to achieve [Webpack tree shaking](https://webpack.js.org/guides/tree-shaking/#components/sidebar/sidebar.jsx).

File: .babelrc
Comment (youknowriad) [inline .babelrc]: Also, es2015 is included in latest, I think it's not needed here

**Comment by nylen** [comment]
This now has `mocha` unit tests :tada:

A few notes about the test setup:

- I ignored `yarn.lock`, but maybe we should consider committing this file instead?
- I'm using [`mocha-webpack`](https://github.com/zinserjan/mocha-webpack) to run the tests through the same webpack transformations as the actual build.  It's pretty important to keep our build and our test environment in sync like this - it's something that's currently causing us problems in Calypso (https://github.com/Automattic/wp-calypso/issues/12203).  See this article for details, configuration explanations, and next steps for getting testing working in a browser:  http://randycoulman.com/blog/2016/03/22/testing-with-mocha-and-webpack/
- I added `dirty-chai` because I [don't really trust property-based assertions](https://github.com/prodatakey/dirty-chai#why).  It's easy to make a typo like `expect( x ).to.be.ture;` - this will appear to succeed but will not actually do anything.

File: .babelrc
Comment (nylen) [inline .babelrc]: Hm, interesting.  At one point I was getting an error until I removed this (http://stackoverflow.com/a/42834190) but now it is working fine.

File: .babelrc
Comment (nylen) [inline .babelrc]: Is there a further change we should make here?  I'm not sure if you can do `latest: { modules: false }` for example, or how to test whether this is working as intended.

File: test/blocks/index.js
Comment (youknowriad) [inline test/blocks/index.js]: Is there a reason why you would use * instead of explicit imports?

File: test/blocks/index.js
Comment (youknowriad) [inline test/blocks/index.js]: Should we use the `/** External/Internal dependecies **` comments (similar to Calypso)?

Comment (nylen) [inline test/blocks/index.js]: Yep, probably :)

File: test/blocks/index.js
Comment (nylen) [inline test/blocks/index.js]: I have no preference either way.

File: .babelrc
Comment (youknowriad) [inline .babelrc]: No, the last change is good 👍 

Yes, hard to know if it's working correctly. I guess we'll see when we import some function for `lodash`. We could check that it's not including the whole library

**Comment by nylen** [comment]
In e157c9a I changed `name` to `slug`.  So far, we're really dealing with the internal ID of the block, which I am calling a slug because that's the WordPress term for a dashed identifier, including a plugin's ID.

I'm open to suggestions here, but I don't think it should be called `name`.  The name is the human-readable block name, which doesn't exist yet but will be different and probably not necessarily unique.

**Comment by youknowriad** [comment]
> I ignored yarn.lock, but maybe we should consider committing this file instead?

On the initial build PR. I've used yarn as well but if we commit the `yarn.lock`, all contributors should use `yarn` to avoid making `yarn.lock` out of sync when adding dependencies. @aduth raised a good point, that this would add a new thing to learn for contributors, and we may want to stick with npm for now

File: test/blocks/index.js
Comment (youknowriad) [inline test/blocks/index.js]: In this particular case, it's not making any difference, but worth noting that in other use cases when importing from external libraries, it can have an impact on `tree shaking` 

```js
import * as _ from 'lodash'; console.log( _.uniqueId() ); // This will include the whole lodash
import { uniqueId } from 'lodash'; console.log( uniqueId() ); // This will include only uniqueId
```

**Comment by youknowriad** [comment]
This is looking solid for me. Thanks, @nylen for the test infra. Waiting for a 👍 to merge

File: test/blocks/index.js
Comment (aduth) [inline test/blocks/index.js]: If we're going to pull in and assume plugins to be used consistently, could we somehow apply it as part of the overall test initialization?

File: test/blocks/index.js
Comment (aduth) [inline test/blocks/index.js]: Would `blocks.unregisterBlock` be a reasonable reasonable to work around this problem?

File: webpack.config.test.js
Comment (aduth) [inline webpack.config.test.js]: I see it in `mocha-webpack`'s docs, and even upon reading the docs of `webpack-node-externals` itself, it's still not clear why this is necessary. Does it make the build more performant? Is it any different than:

```js
resolve: {
	modules: baseConfig.resolve.modules.filter( ( source ) => source !== 'node_modules' )
}
```

File: package.json
Comment (aduth) [inline package.json]: Seeing that `mocha-webpack` has a `--webpack-config` option and our Webpack config already includes some environment-specific handling, do you have any opinion on whether it might make sense to merge `webpack.config.test.js` into the base config and change this to:

```
"test": "NODE_ENV=test mocha-webpack --webpack-config webpack.config.js test/"
```

Trying to think of ways to avoid root directory becoming too sprawling with files.

**Review by aduth — COMMENTED** [review]
Did you look to see if there's a chance we could integrate QUnit without too much trouble? Seems like it will suffer many of the same gotchas as Mocha, but if we could pull it off would certainly make eventual integration a fair bit less... complicated.

File: webpack.config.test.js
Comment (youknowriad) [inline webpack.config.test.js]: I tried without it and using your alternative. I don't understand all what's happening here but it's not working without this.

**Comment by youknowriad** [comment]
I took the liberty to address some of @aduth feedback on the last commit. I rebased to avoid two npm test scripts.

File: test/blocks/index.js
Comment (youknowriad) [inline test/blocks/index.js]: Could we consider that this is already tested, since we're calling it above and checking the results in the `getBlockSettings` test

**Comment by youknowriad** [comment]
I've rebased this after #295 merge. Also, I had to make a small change to the `webpack.config.js` to allow importing modules in unit tests see a7c6660 @aduth

File: webpack.config.js
Comment (aduth) [inline webpack.config.js]: Maybe a preferred style, but just FYI this would have still worked fine if left as it was.

File: webpack.config.js
Comment (aduth) [inline webpack.config.js]: Not sure what the overhead is like for requiring a module, probably not much. But an option may be to inline it as:

```js
config.externals = [ require( 'webpack-node-externals' )() ];
```

I'm not too concerned about it.

File: modules/blocks/index.js
Comment (aduth) [inline modules/blocks/index.js]: Semicolon.

File: modules/blocks/index.js
Comment (aduth) [inline modules/blocks/index.js]: Minor: I find the pattern of calling a function for the sole purpose of maybe throwing an error a little odd. Another option might be:

```js
if ( ! isValidBlockSlug( slug ) ) {
	throw new TypeError( 'Block slugs must contain a namespace prefix. Example: my-plugin/my-custom-block' );
}
```

**Review by aduth — DISMISSED** [review]
There's now an error when running `npm run build` about `assets/stylesheets/main.scss` . We'll probably need to add each "submodule" as a root of `modules`. Let me poke around at at that.

File: webpack.config.js
Comment (youknowriad) [inline webpack.config.js]: Yeah, I know but I don't know, this seems more "logic" to me (obviously biased)

**Comment by aduth** [comment]
>There's now an error when running `npm run build`

This is now fixed in 59d3ea0, but still appears to be some issues with test assuming `wp.blocks` globals. Might need to inject those or apply an externals mask like the one previously removed in 8b866fa02ed04443c11910c49005646b58d20718 in test context?

**Comment by aduth** [comment]
Oh I just noticed now; do we want tests in root directory, in each "submodule"'s root directory, or nested as a subdirectory of the file under test?

**Comment by youknowriad** [comment]
I do have a preference for "nested as a subdirectory of the file under test" (Closer to the source)

File: modules/blocks/index.js
Comment (nylen) [inline modules/blocks/index.js]: The reason for this design was so that the `validateBlockSlug` function could choose from multiple possible errors to throw, and so that this functionality could be tested independently of registering a block.

File: test/blocks/index.js
Comment (nylen) [inline test/blocks/index.js]: I'm not really a fan of the approach though, it should be tested explicitly.  I think @aduth's suggestion `unregisterBlock` is a decent improvement.

**Comment by aduth** [comment]
Issue with `mocha-webpack` is that it won't respect our custom `entry` definitions:

>Maybe you noticed that `entry`, `output.filename` and `output.path` are completely missing in this config. mocha-webpack does this automatically for you and if you would specify it anyway, it will be overridden by mocha-webpack.

https://www.npmjs.com/package/mocha-webpack

**Comment by nylen** [comment]
Thanks, @youknowriad and @aduth for reviewing.

In a512271 I added `blocks.unregisterBlock`, which allowed for some improvements to the tests and for getting rid of the separate function for `validateBlockSlug`.

I'm not having any problems with `npm test` or `npm run build`, and I don't follow the globals issues mentioned in https://github.com/WordPress/gutenberg/pull/289#issuecomment-287801878.  Is there anything left to do here?

One remaining question here is whether to use QUnit or Mocha for test infrastructure.  I find QUnit very awkward to work with.

In terms of eventual integration with WP core, in the worst case this is going to mean that we add a separate Travis CI job to run Mocha tests.  I think this is OK, because there will be lots of tests for this project, so they will take a while to run :wink: 

I think the added developer convenience of using Mocha is well worth it, and maybe we can also start a new standard for core JS tests.

So, in short, IMO this is ready to merge so we can move forward.

File: modules/blocks/index.js
Comment (nylen) [inline modules/blocks/index.js]: As of a512271 this is gone - I agree it's cleaner to just inline this logic into `registerBlock`.

**Comment by youknowriad** [comment]
I don't think it's related to this PR, but we're having errors like "TypeError: n(...) is not a function" on the plugin page. Seems like a conflict between the webpack build and some WordPress global variable? Any idea about this?

**Comment by nylen** [comment]
I think those errors are related to this PR.  Here are the ones I'm seeing (and none in `master` after a rebuild):

2017-03-22t09 30 23-0300

Seems like we need to add some integration tests for the plugin page as well.

**Comment by youknowriad** [comment]
I found that babel `transform-runtime` plugin is causing these errors, but I can't find why exactly.

**Comment by youknowriad** [comment]
Replacing `transform-runtime` with `babel-polyfill` fixes the bug, but we can not include the polyfill in the three modules (duplicate), I'm thinking of adding the polyfill as an external script tag. Thoughts?

**Comment by aduth** [comment]
65066e1 fixes errors from Webpack.

There's another error from `registerBlock` for how we're using old signature for the text block. Throwing errors crashes the entire editor initialization. Maybe a little heavy handed? Perhaps we can consider `console.error` instead.

**Comment by youknowriad** [comment]
Thanks for the fix @aduth I fixed the `npm test` command to correctly check lint errors and fix the remaining lint issues in b95c775

**Comment by youknowriad** [comment]
I fixed the last `registerBlock` error in e5f72ca and the React minification warning. I'm rebasing (too many junk commits) and merging. Tks @aduth and @nylen. We can make changes to these on follow-up PRs if needed

**Comment by aduth** [comment]
I'd still like some feedback on my earlier point:

>Throwing errors crashes the entire editor initialization. Maybe a little heavy handed? Perhaps we can consider `console.error` instead.

**Comment by youknowriad** [comment]
@aduth I agree, We shouldn't crash on wrong blocks, but just ignore them. 

👍 for console.error for me

**Pull request merged**
